subprojects/introspect/source/jcli/introspect/groups.d

```diff
@@ -0,0 +1,332 @@
+        }
+
+        // TODO: is bit array worth it?
+        bool[Types.length] isNotRootCache;
+        foreach (parentIndex, children; childrenGraph)
+        {
+            foreach (childNode; children)
+                isNotRootCache[childNode.typeIndex] = true;
+        }
+
+        {
+            // Check for cicles in the graph.
+            // Cicles will stifle the compiler (but I'm not sure).
+            // TODO: use bit array?
+            bool[Types.length] visited = false;
+
+            bool isCyclic(size_t index)
+            {
+                if (visited[index])
+                    return true;
+                
+                visited[index] = true;
+                foreach (childNode; childrenGraph[index])
+                {
+                    if (isCyclic(childNode.typeIndex))
+                        return true;
+                }
+                return false;
+            }
+
+            foreach (parentIndex, isNotRoot; isNotRootCache)
+            {
+                if (isNotRoot)
+                    continue;
+                visited[] = false;
+                if (isCyclic(parentIndex))
+                {
+                    // TODO: report more info here.
+                    assert(false, "The command graph contains a cycle");
+                }
+            }
+        }
+		
+        import std.array : appender;
+        import std.format : formattedWrite;
+        import std.algorithm : map;
+
+        auto ret = appender!string;
+
+        {
+            ret ~= "template Mappings() {";
+            foreach (key, index; typeToIndex)
+                formattedWrite(ret, "enum int %s = %d;", key, index);
+            ret ~= "}";
+        }
+
+        {
+            // auto rootTypes = appender!string;
+            // rootTypes ~= "alias RootTypes = AliasSeq!(";
+
+            auto rootTypeIndices = appender!string;
+            rootTypeIndices ~= "immutable int[] rootTypeIndices = [";
+            {
+                size_t appendedCount = 0;
+                foreach (nodeIndex, isNotRoot; isNotRootCache)
+                {
+                    if (isNotRoot)
+                        continue;
+                    if (appendedCount > 0)
+                    {
+                        rootTypeIndices ~= ", ";
+                        // rootTypes ~= ", ";
+                    }
+                    appendedCount++;
+                    
+                    formattedWrite(rootTypeIndices, "%d", nodeIndex);
+                    // formattedWrite(rootTypes, "Types[%d]", nodeIndex);
+                }
+            }
+            rootTypeIndices ~= "];\n";
+
+            ret ~= rootTypeIndices[];
+            ret ~= "\n";
+
+            // rootTypes ~= ");\n";
+            // ret ~= rootTypes[];
+            // ret ~= "\n";
+        }
+        {
+            formattedWrite(ret, "immutable Node[][] Adjacencies = %s;\n", childrenGraph);
+        }
+
+        // Extra info, not actually used
+        static if (0)
```

Comment:
Is this for debugging purposes? Otherwise it might be wise to consider if we should just remove it instead of disable it.

Reply:
We'll keep it for now then, and maybe remove it once we're certain the code is in a more stable state (i.e. we won't be working on it much more).